### ISSUE TITLE
API configuration is poorly documented : ClientResponseException

### ISSUE THREAD

**Issue by ssbarnea** [jenkinsci/openstack-cloud-plugin#139]
It seems that even if support for v3 OS APi was added via https://github.com/jenkinsci/openstack-cloud-plugin/pull/133 users will continue to strugle trying to configure the plugin (2.17) due to lack of documentation and less that useful error messages.

I do know that this instance is working with both v2 and v3 API but still I was not able to find any way to configure the plugin so it will past the testing phase.
```
set|grep -a OS_
OS_AUTH_URL=http://10.8.188.11:5000/v3/
OS_IDENTITY_API_VERSION=3
OS_PASSWORD=secret123
OS_PROJECT_DOMAIN_NAME=Default
OS_PROJECT_NAME=johnd
OS_USERNAME=johnd
OS_USER_DOMAIN_NAME=Default
```
As you can see the help section gives example only for the v2 syntax and does this using the [obsolete naming](https://wiki.openstack.org/wiki/OpenStackClient/Authentication).

![info](http://s3.sbarnea.com/ss/Configure_System_Jenkins_2017-02-07_09-14-38.png)

I had two approaches, both failing:

* v3: Identify: `johnd:johnd:Default` and the endpoint ending in /v3/
* v2: Identify: `johnd:johnd` and the endpoint ending in /v2/

On both I got:
```
Cannot connect to specified cloud, please check the identity and credentials: Not Found
ClientResponseException{message=Not Found, status=404, status-code=NOT_FOUND}
	at org.openstack4j.core.transport.HttpExceptionHandler.mapException(HttpExceptionHandler.java:38)
```

**Comment by pjdarton** [comment]
If you're using plugin version 2.17 then you are *not* running a version that contains https://github.com/jenkinsci/openstack-cloud-plugin/pull/133
While that pull-request (adding v3 support) has been merged into the master branch, and that *will* (in the near future) be used to build version 2.18, there is currently no official release that contains support for v3-auth.
i.e. If you want support for v3 auth then you will need to EITHER build your own version of the plugin, or download one from https://jenkins.ci.cloudbees.com/job/plugins/job/openstack-cloud-plugin/ (just beware that the cloudbees site is building various peoples' pull-requests, and the binaries from PR#133 are now long gone - you may find that https://jenkins.ci.cloudbees.com/job/plugins/job/openstack-cloud-plugin/312/org.jenkins-ci.plugins$openstack-cloud/artifact/org.jenkins-ci.plugins/openstack-cloud/2.18-SNAPSHOT/openstack-cloud-2.18-SNAPSHOT.hpi works for you though).
If you get really stuck, let me know an email address and I can send you a copy of the one I'm using.

Regarding the 404, I'm pretty sure that a 404 isn't directly related to the plugin code.  I suspect that this is more likely because you've not given the plugin the "right" endpoint URL.
Finding out exactly what the "right" URL is can be non-trivial.  In theory, "your local OpenStack deployment administrator can tell you this" ... but you are your own OpenStack admin then you need to find this out somehow :-/
This isn't the sort of thing that the plugin can simply "discover" through software alone - it very much relies on the user telling it where to start.  If the user doesn't know where they deployed the Identity service to then the plugin software can't tell them.
This issue is further compounded by there only being limited support for v3-auth right now.  e.g. Where I work, we run "Horizon" as a web front end for interactive use (point browser at http://somehostname/, login, get pretty WebUI showing an overview of OpenStack deployment) and, if I use that to look at "Access & Security -> API Access -> Identity", it correctly lists the correct endpoint for our v2-auth cloud deployment (it says that the Identify service is at "http://somehostname:5000/v2.0").  However, if I do the same on our v3-auth cloud then it gives the wrong answer (Horizon tells me my Identity service is at "https://otherhostname:5000/v2.0" whereas I need to tell Jenkins to use "https://otherhostname:5000/v3"); I don't know if this is a bug in Horizon or an error in my deployment.

Regarding the OS_AUTH_URL environment variables etc, I suspect that those aren't coming from the plugin code - I did a string search for a few of them and they're not in the plugin source code.  I think that it's rather more likely that its your own cloud-init code (that's being run within the VM on OpenStack when the VM boots up) that's setting those.


As for documentation, you will (I hope) be pleased to hear that the help text in the next version of the plugin does tell you to use PROJECT_NAME:USER_NAME:DOMAIN_NAME syntax for v3 auth: https://github.com/jenkinsci/openstack-cloud-plugin/blob/master/src/main/resources/jenkins/plugins/openstack/compute/JCloudsCloud/help-identity.html
If, when you look at the plugin help, you only see the "old" syntax that you mentioned then the plugin you have installed in your Jenkins doesn't support v3 auth and you need to get a newer version (i.e. newer than 2.17).

Note: If you can provide better help text for the endpoint URL field (which is currently just https://github.com/jenkinsci/openstack-cloud-plugin/blob/master/src/main/resources/jenkins/plugins/openstack/compute/JCloudsCloud/help-endPointUrl.html ) that can make it easier for others to work out what their URL should be then please do contribute and it can be incorporated into future versions of the plugin (I'm strongly in favor of help-text that really helps)

**Comment by ssbarnea** [comment]
@pjdarton Thank you. In fact after compiling the main branch myself I was able to make it authenticate. The problem was that I was trying to use the last release which didn't include your changes. 

I had to postpone my test of this plugin due to other priorities but I will be back.

**Comment by ssbarnea** [comment]
I am closing this because this was fixed in trunk.

**Comment by olivergondza** [comment]
The feature is released in 2.18. Closing this.
